File: NeetCode/Two_Pointers/Container_with_most_water/container_with_most_water.py
from typing import Optional, List


class Solution:
    def maxArea(self, heights: List[int]) -> int:
        left = 0
        right = len(heights) - 1

        mArea = 0
        mLeft, mRight = 0, 0

        while left < right:
            mLeft = max(mLeft, heights[left])
            mRight = max(mRight, heights[right])
            volume = min(mLeft, mRight) * (right - left)
            mArea = max(mArea, volume)
            if heights[left] < heights[right]:
                left += 1
            else:
                right -= 1

        return mArea


if __name__ == "__main__":
    sol = Solution()
    height = [1, 7, 2, 5, 4, 7, 3, 6]

    print(sol.maxArea(height))
    print("Running Solution...")

File: NeetCode/Two_Pointers/Container_with_most_water/Container_with_most_water.cpp
// Container_with_most_water.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int maxArea(vector<int> &heights)
    {
        int left{0}, maxSeen{0}, curr{0}, lVal{0}, rVal{0};
        int right = heights.size() - 1;

        while (left < right)
        {
            lVal = heights[left];
            rVal = heights[right];
            curr = (right - left) * min(lVal, rVal);
            maxSeen = max(maxSeen, curr);

            if (lVal <= rVal)
            {
                left++;
            }
            else
            {
                right--;
            }
        }
        return maxSeen;
    }
};
